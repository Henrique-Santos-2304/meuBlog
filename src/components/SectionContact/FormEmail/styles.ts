import styled, { css } from "styled-components";

export const Wrapper = styled.div`
  position: relative;
`;
export const ModalForm = styled.div`
  width: max-content;
  position: absolute;
  top: -1rem;
  left: 50%;
  padding: 0.6rem 1.4rem;
  transform: translateX(-50%);
  font-size: 1.4rem;
  background: rgba(255, 255, 255, 0.15);
  color: ${({ theme }) => theme.colors.effectsPrimary};
  text-align: center;
  border-radius: 0.5rem;
`;
export const Form = styled.form`
  padding: 3rem 0.5rem;
  margin: 0 auto;
  width: 80%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

export const Inputs = styled.input`
  ${({ theme }) => css`
    padding: ${theme.spacings.xxsmall};
    border-bottom: 0.1rem solid ${theme.colors.secondary};
    font-size: 0.8rem;
    background: transparent;
    border-radius: 0.7rem;
    text-align: start;
    color: ${theme.colors.primary};
    transition: all 0.3s linear;

    ::placeholder {
      color: rgba(0, 0, 0, 0.45);
    }
    &:hover {
      transform: translateY(-3%);
    }
    &:hover,
    &:focus {
      background: rgba(255, 255, 255, 0.1);
      border: 0.1rem solid ${theme.colors.effectsPrimary};
    }
  `}
`;

export const TextAreas = styled.textarea.attrs({
  cols: 21,
  rows: 6,
})`
  ${({ theme }) => css`
    padding: ${theme.spacings.xsmall};
    overflow: hidden;
    border: 0;
    background: rgba(0, 0, 0, 0.1);
    border-radius: 0.6rem;
    transition: all 0.3s linear;
    color: ${theme.colors.primary};
    &:hover {
      transform: translateY(-3%);
    }
    ::placeholder {
      color: rgba(0, 0, 0, 0.45);
    }
    &:focus {
      border: 0.1rem solid ${theme.colors.effectsPrimary};
      outline: none;
    }
  `}
`;
export const Button = styled.button`
  ${({ theme }) => css`
    padding: 0.5rem ${theme.spacings.xsmall};
    background: transparent;
    color: ${theme.colors.effectsPrimary};
    font-weight: 700;
    border-radius: 0.5rem;
    border: 0.1rem solid ${theme.colors.effectsPrimary};
  `}
`;
