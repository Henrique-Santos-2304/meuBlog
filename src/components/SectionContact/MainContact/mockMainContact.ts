export const mockMainContact = {
  dataContact: {
    title: "contato",
    ariaLabel: "contato",
    ariaLabelTitle: "Titulo da seção",
    Input: {
      name: "name",
      placeholder: "Digite seu Nome",
      type: "text",
      msgError: "Digite um nome",
      textLabel: "Nome",
    },
    inputEmail: {
      name: "email",
      placeholder: "Digite seu email",
      type: "email",
      msgError: "Digite um email",
      textLabel: "Email",
    },
    inputMessage: {
      name: "message",
      placeholder: "Digite sua mensagem",
      type: "text",
      msgError: "Digite uma mensagem",
      textLabel: "Mensagem",
    },
    buttonSend: {
      text: "enviar",
      name: "button",
      type: "submit",
    },
    msgSendEmail: {
      msgSucess: "Enviado com sucesso",
      msgError: "Falha no envio",
      ariaLabel: "enviar email",
      msgFormatEmail: "Formato de email inválido",
    },
  },
  formEmailData: {
    services: "string;",
    form: "string;",
    user: "string;",
  },
};
