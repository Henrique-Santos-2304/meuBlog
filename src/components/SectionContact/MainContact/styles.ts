import styled, { css } from "styled-components";
export const Wrapper = styled.section`
  ${({ theme }) => css`
    max-width: ${theme.grid.containerContent};
    margin: 0 auto;
    padding-top: 0.5rem;

    h1 {
      text-align: center;
    }
    @media (min-width: ${theme.media.medium}) {
      padding: 2rem ${theme.spacings.small};
    }
    @media (min-width: ${theme.media.xmedium}) {
      padding: 2rem ${theme.spacings.large};
    }
  `}
`;
