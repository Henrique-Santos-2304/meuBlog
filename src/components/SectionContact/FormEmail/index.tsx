/* eslint-disable @typescript-eslint/no-non-null-assertion */
import * as S from "./styles";
import { useEffect, useRef, useState } from "react";
import { useForm } from "react-hook-form";
import emailjs from "emailjs-com";
import ContentInput from "../ContentInput";
import { contact, formDatas } from "graphql/typesQueries/types";

type propsContact = {
  dataContact: contact;
  formEmailData: formDatas;
};

const FormEmail = ({ dataContact, formEmailData }: propsContact) => {
  const {
    handleSubmit,
    register,
    formState: { errors },
  } = useForm();
  const formRef = useRef<HTMLFormElement>(null);
  const [msgForm, setMsgForm] = useState("");

  const onSubmit = async () => {
    await emailjs
      .sendForm(
        formEmailData.services,
        formEmailData.form,
        formRef.current!,
        formEmailData.user
      )
      .then(
        function () {
          setMsgForm(dataContact.msgSendEmail.msgSucess);
        },
        function () {
          setMsgForm(dataContact.msgSendEmail.msgError);
        }
      );
  };
  useEffect(() => {
    setTimeout(() => {
      setMsgForm("");
    }, 10000);
  }, [msgForm]);

  return (
    <S.Wrapper>
      {msgForm && (
        <S.ModalForm aria-label={dataContact.msgSendEmail.ariaLabel}>
          {msgForm}
        </S.ModalForm>
      )}
      <S.Form onSubmit={handleSubmit(onSubmit)} ref={formRef}>
        {/* Input de Nome */}
        <ContentInput
          error={errors.name && errors.name.message}
          typeLabel={dataContact.Input.name}
          txtLabel={dataContact.Input.textLabel}
        >
          <S.Inputs
            id={dataContact.Input.name}
            type={dataContact.Input.type}
            placeholder={dataContact.Input.placeholder}
            {...register(dataContact.Input.name, {
              required: dataContact.Input.msgError,
            })}
          />
        </ContentInput>
        {/* Input de Email */}
        <ContentInput
          error={errors.email && errors.email.message}
          typeLabel={dataContact.inputEmail.name}
          txtLabel={dataContact.inputEmail.textLabel}
        >
          <S.Inputs
            id={dataContact.inputEmail.name}
            type={dataContact.inputEmail.type}
            placeholder={dataContact.inputEmail.placeholder}
            {...register(dataContact.inputEmail.name, {
              required: dataContact.inputEmail.msgError,
              pattern: {
                value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i,
                message: dataContact.msgSendEmail.msgFormatEmail,
              },
            })}
          />
        </ContentInput>

        {/* Input de Mensagem  */}
        <ContentInput
          error={errors.message && errors.message.message}
          typeLabel={dataContact.inputMessage.name}
          txtLabel={dataContact.inputMessage.textLabel}
        >
          <S.TextAreas
            id={dataContact.inputMessage.name}
            placeholder={dataContact.inputMessage.placeholder}
            {...register(dataContact.inputMessage.name, {
              required: dataContact.inputMessage.msgError,
            })}
          />
        </ContentInput>

        <S.Button
          type={dataContact.buttonSend.type}
          value={dataContact.buttonSend.text}
        />
      </S.Form>
    </S.Wrapper>
  );
};

export default FormEmail;
