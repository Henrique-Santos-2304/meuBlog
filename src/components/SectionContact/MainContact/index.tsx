import Headings from "components/globalComponents/Headings";
import { contact, formDatas } from "graphql/typesQueries/types";
import FormEmail from "../FormEmail";
import * as S from "./styles";

type propsContact = {
  dataContact: contact;
  formEmailData: formDatas;
};

const MainContact = ({ dataContact, formEmailData }: propsContact) => (
  <>
    <S.Wrapper id="contacts">
      <Headings
        msg={dataContact.title}
        fontSizeText="2rem"
        marginText="0 0 0 3rem"
        aria-label={dataContact.ariaLabelTitle}
      />
      <FormEmail dataContact={dataContact} formEmailData={formEmailData} />
    </S.Wrapper>
  </>
);

export default MainContact;
