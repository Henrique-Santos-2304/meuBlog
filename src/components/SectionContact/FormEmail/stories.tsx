import { Story, Meta } from "@storybook/react";
import FormEmail from ".";
import { mockMainContact } from "../MainContact/mockMainContact";

export default {
  title: "FormEmail",
  component: FormEmail,
} as Meta;

export const Basic: Story = () => <FormEmail {...mockMainContact} />;

export const Default: Story = () => <FormEmail {...mockMainContact} />;
