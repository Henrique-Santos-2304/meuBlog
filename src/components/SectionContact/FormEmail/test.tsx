import FormEmail from ".";
import { act, fireEvent, screen } from "@testing-library/react";
import { renderTheme } from "utils/testRenderTheme";
import { mocked } from "ts-jest/utils";
import emailjs from "emailjs-com";
import { mockMainContact } from "../MainContact/mockMainContact";

jest.mock("emailjs-com");

describe("<FormEmail />", () => {
  it("should render ", () => {
    const { container } = renderTheme(<FormEmail {...mockMainContact} />);
    expect(container).toMatchSnapshot();
  });
  it("all fields of the imaginary form on the screen", () => {
    renderTheme(<FormEmail {...mockMainContact} />);
    const labelName = screen.getByText(/nome/i);
    const labelEmail = screen.getByText(/email/i);
    const labelMessage = screen.getByText(/mensagem/i);

    const inputName = screen.getByRole("textbox", {
      name: /nome/i,
    });
    const inputEmail = screen.getByRole("textbox", {
      name: /email/i,
    });
    const inputMessage = screen.getByRole("textbox", {
      name: /mensagem/i,
    });

    const buttonSubmit = screen.getByRole("button", { name: /enviar/i });
    expect(labelName).toBeInTheDocument();
    expect(labelEmail).toBeInTheDocument();
    expect(labelMessage).toBeInTheDocument();

    expect(inputName).toBeInTheDocument();
    expect(inputEmail).toBeInTheDocument();
    expect(inputMessage).toBeInTheDocument();

    expect(buttonSubmit).toBeInTheDocument();
  });
  it("should Have 3 alerts with 3 invalid fields", async () => {
    renderTheme(<FormEmail {...mockMainContact} />);
    const button = screen.getByRole("button", { name: /enviar/i });

    fireEvent.submit(button);
    expect(await screen.findAllByRole("alert")).toHaveLength(3);
  });

  it("should have sent message Sucess when the data is valid", async () => {
    jest.useFakeTimers();
    const mockEmaijs = mocked(emailjs);

    renderTheme(<FormEmail {...mockMainContact} />);
    fireEvent.input(screen.getByRole("textbox", { name: /nome/i }), {
      target: {
        value: "test",
      },
    });
    fireEvent.input(screen.getByRole("textbox", { name: /email/i }), {
      target: {
        value: "test@mail.com",
      },
    });

    fireEvent.input(screen.getByRole("textbox", { name: /mensagem/i }), {
      target: {
        value: "Mensagem",
      },
    });

    fireEvent.submit(screen.getByRole("button", { name: /enviar/i }));
    expect(await screen.queryAllByRole("alert")).toHaveLength(0);
    await mockEmaijs.sendForm.mockResolvedValueOnce({
      text: "OK",
      status: 200,
    });

    expect(await screen.findByText(/Enviado com sucesso/i)).toBeInTheDocument();
    act(() => {
      jest.runAllTimers(); // trigger setTimeout
    });

    expect(screen.queryByText(/Enviado com sucesso/i)).not.toBeInTheDocument();
  });
  it("should have sent message error when the data is invalid", async () => {
    jest.useFakeTimers();
    const mockEmaijs = mocked(emailjs);

    renderTheme(<FormEmail {...mockMainContact} />);

    fireEvent.input(screen.getByRole("textbox", { name: /nome/i }), {
      target: {
        value: "test",
      },
    });
    fireEvent.input(screen.getByRole("textbox", { name: /email/i }), {
      target: {
        value: "test@mail.com",
      },
    });

    fireEvent.input(screen.getByRole("textbox", { name: /mensagem/i }), {
      target: {
        value: "Mensagem",
      },
    });

    fireEvent.submit(screen.getByRole("button", { name: /enviar/i }));

    mockEmaijs.sendForm.mockRejectedValueOnce({ Error });

    expect(await screen.findByText(/Falha no envio/i)).toBeInTheDocument();

    act(() => {
      jest.runAllTimers(); // trigger setTimeout
    });
    expect(await screen.queryByText(/Falha no envio/i)).not.toBeInTheDocument();
  });
});
