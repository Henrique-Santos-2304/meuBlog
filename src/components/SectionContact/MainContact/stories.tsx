import { Story, Meta } from "@storybook/react";
import MainContact from ".";
import { mockMainContact } from "./mockMainContact";

export default {
  title: "MainContact",
  component: MainContact,
} as Meta;

export const Basic: Story = () => <MainContact {...mockMainContact} />;

export const Default: Story = () => <MainContact {...mockMainContact} />;
