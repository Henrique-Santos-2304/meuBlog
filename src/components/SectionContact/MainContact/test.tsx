import MainContact from ".";
import { screen } from "@testing-library/react";
import { renderTheme } from "utils/testRenderTheme";
import { mockMainContact } from "./mockMainContact";

describe("<MainContact />", () => {
  it("should render ", () => {
    const { container } = renderTheme(<MainContact {...mockMainContact} />);
    expect(container).toMatchSnapshot();
  });
  it("should render ", () => {
    renderTheme(<MainContact {...mockMainContact} />);
    const title = screen.getByRole("heading", {
      name: /contato/i,
    });
    expect(title).toBeInTheDocument();
  });
});
