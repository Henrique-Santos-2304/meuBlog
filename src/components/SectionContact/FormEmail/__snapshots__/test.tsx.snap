// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormEmail /> should render  1`] = `
.c0 {
  position: relative;
  max-width: 74rem;
}

.c1 {
  padding: 0.3rem 0.5rem;
  margin: 0rem auto;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c4 {
  padding: 0.8rem 1.6rem;
  border-bottom: 0.1rem solid #1e1e1e;
  font-size: 1.4rem;
  background: transparent;
  border-radius: 0.7rem;
  text-align: start;
  color: #f8f8f8;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
}

.c4::-webkit-input-placeholder {
  color: rgba(0,0,0,0.45);
}

.c4::-moz-placeholder {
  color: rgba(0,0,0,0.45);
}

.c4:-ms-input-placeholder {
  color: rgba(0,0,0,0.45);
}

.c4::placeholder {
  color: rgba(0,0,0,0.45);
}

.c4:hover,
.c4:focus {
  background: rgba(255,255,255,0.1);
  border: 0.1rem solid #b7f600;
}

.c5 {
  padding: 0.8rem;
  overflow: hidden;
  border: 0;
  font-size: 1.4rem;
  font-family: 'Poppins',sans-serif;
  background: rgba(0,0,0,0.1);
  border-radius: 0.6rem;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
  color: #f8f8f8;
}

.c5::-webkit-input-placeholder {
  color: rgba(0,0,0,0.45);
}

.c5::-moz-placeholder {
  color: rgba(0,0,0,0.45);
}

.c5:-ms-input-placeholder {
  color: rgba(0,0,0,0.45);
}

.c5::placeholder {
  color: rgba(0,0,0,0.45);
}

.c5:hover,
.c5:focus {
  border: 0.1rem solid #b7f600;
  outline: none;
}

.c6 {
  margin-bottom: 0.3rem;
  padding: 0.5rem 1.6rem;
  background: transparent;
  color: #b7f600;
  font-weight: 700;
  border-radius: 0.5rem;
  border: 0.1rem solid #b7f600;
  -webkit-transition: all 0.4s ease-in-out;
  transition: all 0.4s ease-in-out;
}

.c6:hover {
  color: #f8f8f8;
  border: 0.1rem solid #f8f8f8;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 1rem 0.6rem;
  margin: 0.6rem 0;
}

.c3 {
  font-size: 1.2rem;
  margin: 0.4rem;
  color: #f8f8f8;
}

@media (min-width:620px) {
  .c4 {
    width: 24rem;
  }
}

@media (min-width:768px) {
  .c4 {
    font-size: 1.4rem;
  }
}

@media (min-width:620px) {
  .c5 {
    width: 24rem;
  }
}

@media (min-width:768px) {
  .c5 {
    font-size: 1.4rem;
  }
}

@media (min-width:768px) {
  .c3 {
    font-size: 1.4rem;
  }
}

<div>
  <div
    class="c0"
  >
    
    <form
      class="c1"
    >
      <div
        class="c2"
      >
        <label
          class="c3"
          for="name"
        >
          Nome
        </label>
        <input
          class="c4"
          id="name"
          name="name"
          placeholder="Digite seu Nome"
          type="text"
        />
      </div>
      <div
        class="c2"
      >
        <label
          class="c3"
          for="email"
        >
          Email
        </label>
        <input
          class="c4"
          id="email"
          name="email"
          placeholder="Digite seu email"
          type="email"
        />
      </div>
      <div
        class="c2"
      >
        <label
          class="c3"
          for="message"
        >
          Mensagem
        </label>
        <textarea
          class="c5"
          cols="25"
          id="message"
          name="message"
          placeholder="Digite sua mensagem"
          rows="5"
        />
      </div>
      <input
        class="c6"
        type="submit"
        value="enviar"
      />
    </form>
  </div>
</div>
`;
